Refactor builder base

shahzebsiddiqui opened the pull request.

shahzebsiddiqui committed:
add uuid.uuid4 as a sub-directory when formulating the test directory.
There was no need for _get_testdir() method in BuilderBase since we
can retrieve this via self.metadata['testdir']. We we make use
of self.testdir when referencing test directory, which could also
be referenced via self.metadata['testdir'].
We don't need to return testpath from _write_test() since we can
reference class variable. Likewise we don't need to pass testfile
to run_tests() method since we can reference via
self.metadata['testpath']

shahzebsiddiqui committed:
change subdirectory to use the config name and last 8 chars of uuid.

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

shahzebsiddiqui committed:
undo uuid change for self.testdir

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -427,19 +420,14 @@ def run(self):
         return result
 
-    def run_tests(self, testfile):
+    def run_tests(self):
```
So the issue that I have here is that the client (before this change) is flexible to be updated with a different test file to run. That's actually the entire point - you can instantiate it, and then update the file at runtime without needing to start over. I would prefer to keep this ability because I can see it being useful for some custom implementation / use case in the future.

shahzebsiddiqui replied:
I see your point, i didn't realize you intended to import the BuilderBase class directly. It makes sense, ill fix this.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -449,11 +437,12 @@ def run_tests(self, testfile):
 
         # Run the test file using the shell
-        cmd = [self.get_shell(), testfile]
+        cmd = [self.get_shell(), self.metadata["testpath"]]
```
If you are going to continually reference self.metadata["testpath"], and removing self._get_testpath, a much more elegant solution is to create a property that essentially returns the same thing, e.g.,

```python
@property
def testpath(self, testpath=None):
        """Based on the testfile path, return the testing directory.
           Returns: full path to testing directory
        """
       return self.metadata.get("testpath")
```
You could also combine this sort of thing with being able to parse the original testpath. Anyway, the way you address the issue is up to you, this is a separate suggestion to make reference of the variable much cleaner than self.metadata every time.

shahzebsiddiqui committed:
allow self.run_tests() to pass a testfile, similarly allow
self._write_test to return the testpath. This would be useful
if user invokes the BuilderBase class directly in a python shell and
wants to customize the behavior.

shahzebsiddiqui committed:
applying black

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -539,6 +540,7 @@ def _write_test(self, lines=None):
         self.logger.debug(
             f"Applying permission 755 to {self.metadata['testpath']} so that test can be executed"
         )
+        return self.metadata['testpath']
```
What is this return added for?

shahzebsiddiqui replied:
this was me undoing my change it's already part of ``devel`` https://github.com/buildntest/buildtest/blob/devel/buildtest/buildsystem/base.py#L554

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -274,23 +275,15 @@ def __str__(self):
 
-        testdir = self._get_testdir()
-        create_dir(testdir)
+        create_dir(self.testdir)
```
Didn't you remove this? This needs to be a function (property)

shahzebsiddiqui replied:
Are you referring to the suggestion you mentioned in https://github.com/buildntest/buildtest/pull/283#discussion_r412489796 

Yes i removed ``self._get_testdir()`` it didn't make sense to me initially, If i understand you would like to retain this feature again?

vsoch replied:
Instead of referencing self.metadata["testdir"] I suggested making a @property function so the user can reference `self.testdir`. I think I even wrote it out for you.

shahzebsiddiqui replied:
Yes i got that, i think we have 2 identical ways to retrieve testdir and it's bit confusing why this was the case and i don't like it either. 

1. Class variable self.testdir on line 249 https://github.com/shahzebsiddiqui/buildtest-framework/blob/192fa17b61b96875ce8a491b903d60580ab01bae/buildtest/buildsystem/base.py#L249


2. self.metadata["testpath"], self.metadata["testdir"] on line 398 https://github.com/shahzebsiddiqui/buildtest-framework/blob/refactor_BuilderBase/buildtest/buildsystem/base.py#L398-399

It's bit redundant having self.testdir, self.metadata["testdir"] and i dont see a purpose of having both.

User can still pass a testdir to BuilderBase class and retrieve the test directory via class variable ``testdir``. So is adding a method ``testpath`` going to solve anything. I am fine btw, of referencing ``self.metadata['testpath']`` to keep things simple.

```
a = BuilderBase(name, recipe, testdir=/path/to/testdir)
a.testdir
```

I am not so familiar with @property, so i'd like to understand what is the end-goal with introducing this method.

vsoch replied:
I prefer the first. When was the self.metadata ["testdir"] added? That doesn't make sense.

vsoch replied:
https://www.freecodecamp.org/news/python-property-decorator/
